### PR TITLE
Add spent column to mobile view

### PR DESCRIPTION
**Pull request by Crazypkr1099**

### CODE REVIEW

**Comment by Kidglove57** [comment]
I opened my own file in the preview. The three columns display perfectly on my iPhone 13 Pro using GBP. No truncating of category names or numbers.

just a thought  - may give a problem with those currencies where even small denominations run to many numbers?

**Comment by shall0pass** [comment]
> just a thought - may give a problem with those currencies where even small denominations run to many numbers?

That's a good point. I put a transaction of 2,500,000 in to see what the display looked like.  The numbers do start to run together, though I can't really claim the latest release is too much better in this regard.

Demo.actualbudget.org
Screenshot_20230904-060842-796
Screenshot_20230904-060727-410

**Comment by Kidglove57** [comment]
But then I suspect that the currencies we are referring too give the same problem on mobile in any finance app? And even in the current two column mode?

Personally I would love to see the spent column added. Just need the check on a smaller mobile such as the iPhone Mini?

**Comment by Crazypkr1099** [comment]
ab

This is what it looks like on iphone mini dimensions

**Comment by Crazypkr1099** [comment]
Recent Commit added flex-basis of 80px to prevent issues with "Projected Savings" as shown in the below photo:
ab

**Comment by MatissJanis** [comment]
👋 Would you mind updating the VRT screenshots in order to get the VRT CI job passing? Instructions can be found here: https://github.com/actualbudget/actual/blob/master/packages/desktop-client/README.md

TLDR:
```
E2E_START_URL=https://192.168.0.178:3001 yarn vrt --update-snapshots
```

**Comment by MatissJanis** [comment]
Also: this is iphone SE. Looks pretty broken due to the small screen size. Would it be possible to hide the extra column for such small screens?

Screenshot 2023-09-05 at 21 30 29

File: packages/desktop-client/src/components/budget/MobileBudgetTableSmall.js [hunk 1]
```diff
@@ -0,0 +1,1092 @@
+import React, { Component, memo, PureComponent, useState } from 'react';
```
[the remark this answers is not in the document]
Comment (Crazypkr1099) [reply packages/desktop-client/src/components/budget/MobileBudgetTableSmall.js]: @matiss.ja I saw you comment and I totally agree that doing it that was is awful. The problem is that if I want to make it dynamic, you cannot just add a "if" statement in the mobiletable file. I need to check the viewport width whenever the size changes, meaning that I need to use the ResponsiveProvider code, and if I do that, I need to change everything else. Do you have any suggestions?

I can easily make this work for Iphone SE and below, but it will not be responsive.

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 2]
```diff
@@ -37,6 +37,8 @@ import { AmountInput } from '../util/AmountInput';
 // import { DragDrop, Draggable, Droppable, DragDropHighlight } from './dragdrop';
 import { ListItem, ROW_HEIGHT } from './MobileTable';
 
+const smallVPWidth = document.documentElement.clientWidth < 365;
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:40]: You can use `useResizeObserver` to listen for screen size changes and then show the blocks depending on that size.

Sample: https://github.com/actualbudget/actual/blob/21d5f117eec2e881c0dbc03ff85f02d63fccf3a8/packages/desktop-client/src/components/budget/MonthPicker.js#L43

Comment (MatissJanis) [reply packages/desktop-client/src/components/budget/MobileBudgetTable.js:40]: I think this is my last comment. Once we have this solved I'll update the screenshots and I think we're good to go :)

Comment (Crazypkr1099) [reply packages/desktop-client/src/components/budget/MobileBudgetTable.js:40]: Unfortunately that will not work as its required to be in a function and not a class. As of right now, there is support for iphone SE and I truly don't believe people are going to resize their desktop screen less than 370px. If anyone wants to take over and add dynamic support, they can.

**Comment by joel-jeremy** [comment]
This looks cool. Doesn't look as cramped on my phone but the category names have less space now. Maybe we should add text overflow handling.
Screenshot_20230906-185355

**Comment by Crazypkr1099** [comment]
I agree Joel, but I feel this is a separate issue that we can work on later.

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 3]
```diff
@@ -774,10 +850,25 @@ class BudgetGroups extends Component {
...
+        setSmallVPWidth(true);
+      } else {
+        setSmallVPWidth(false);
+      }
+    };
+
+    window.addEventListener('resize', handleResize);
+
+    return () => {
+      window.removeEventListener('resize', handleResize);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:869]: **💬 suggestion:** ‏this is a great start. But could you move it to the highest level component in this file and then just pass down as a prop in order to reduce the duplicated logic?

Thanks

**Comment by MatissJanis** [comment]
Updated your branch with `master`. This solves the VRT issues. :)

**Comment by Crazypkr1099** [comment]
Hey @MatissJanis , 

Is there anything else you would like me to do for this PR request. I have done the following:
- Added Spent column to mobile view
- Removed Spent column when less than 365 px for support on very small phones
- Added a toggle button to switch between spent/budgeted view for very small phones

Please let me know thank you.

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 4]
```diff
@@ -416,16 +439,52 @@ class TotalsRow extends PureComponent {
...
-            binding={rolloverBudget.groupBudgeted(group.id)}
-            style={{
-              ...styles.smallText,
-              width: 90,
-              fontWeight: '500',
-              textAlign: 'right',
-            }}
-            type="financial"
-          />
+          {!smallVPWidth ? (
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:442]: Similar thoughts as in [the block of ternaries above](https://github.com/actualbudget/actual/pull/1651/files#diff-60e094a44d38fac03fb722c1cf8eb9ddafcf42fbd2d15c0c9f4c3b7d1ec90474R294). Can you consolidate the logic & remove duplication of components?

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 5]
```diff
@@ -289,30 +284,59 @@ class BudgetCategory extends PureComponent {
...
+              name="budgeted"
+              binding={budgeted}
+              editing={editing}
+              style={{ width: 90 }}
+              textStyle={{ ...styles.smallText, textAlign: 'right' }}
+              categoryId={category.id}
+              month={month}
+              onBudgetAction={onBudgetAction}
+            />
+          ) : null}
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:316]:
```suggestion
          {!smallVPWidth || !showBudgetView ? (
            <BudgetCell
              name="budgeted"
              binding={budgeted}
              editing={editing}
              style={{ width: 90 }}
              textStyle={{ ...styles.smallText, textAlign: 'right' }}
              categoryId={category.id}
              month={month}
              onBudgetAction={onBudgetAction}
            />
          ) : null}
```

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 6]
```diff
@@ -815,11 +894,41 @@ export function BudgetTable(props) {
   let currentMonth = monthUtils.currentMonth();
   let format = useFormat();
 
+  const toggleMobileDisplayPref = useSelector(state => {
+    return (
+      (state.prefs.local && state.prefs.local.toggleMobileDisplayPref) || true
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:899]: The `MobileDisplay` in `toggleMobileDisplayPref` is pretty broad and I'm also not sure I follow the "toggle" part. 

If we're directly reading this pref to decide which column to show, perhaps `mobileShowBudgetedColPref`? Then the `useState` variable could match very closely -- `showBudgetedCol`, for instance. 

Alternatively, something like `mobileBudgetFirstColPref` set to a string value `budgeted` / `spent` might be extra easy to follow through the code. 🤔

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 7]
```diff
@@ -815,11 +894,41 @@ export function BudgetTable(props) {
...
+    if (!showBudgetView) {
+      // remember the pref indefinitely
+      dispatch(savePrefs({ toggleMobileDisplayPref: true }));
+    } else {
+      // Set a cookie for 5 minutes
+      let d = new Date();
+      d.setTime(d.getTime() + 1000 * 60 * 5);
+      document.cookie =
+        'toggleMobileDisplayPref=true;path=/;expires=' + d.toUTCString();
+    }
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:921]: I'd say it's safe to drop the `else` altogether and just indefinitely show the column the person picked last. I do think it's a nice touch to remember it for them. 👏🏻

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 8]
```diff
@@ -847,23 +956,80 @@ export function BudgetTable(props) {
             />
           )}
           <View style={{ flex: 1 }} />
-
-          <View style={{ width: 90, justifyContent: 'center' }}>
-            <Label title="BUDGETED" style={{ color: colors.n1 }} />
-            <CellValue
-              binding={reportBudget.totalBudgetedExpense}
-              type="financial"
+          {!smallVPWidth ? (
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:959]: Same deduplication request for the ternaries from here to ln 1032.

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 9]
```diff
@@ -986,6 +1159,34 @@ function BudgetHeader({
         backgroundColor: colors.p5,
       }}
     >
+      {smallVPWidth ? (
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:1162]: Having this up in the header feels a little too cramped for my taste, especially for a screen that's already super narrow. (It's also on the side of the month heading further away from the column that it affects.)

Personally, my intuition is to tap the BUDGETED/SPENT header to change it. For discoverability, maybe add a subtle indicator that would at least suggest something can be done with the header. Something like this maybe? In the order `BUDGETED - SPENT - BALANCE`, the arrow points from BUDGETED toward where SPENT would be, and vice versa. (I haven't removed the header toggle text button here, but that would go.)

budgeted-spent


CSS for the div surrounding the BUDGETED/SPENT text is roughly:
```css
// padding + negative margin to keep the column text aligned
padding: '0 8px';
margin: '0 -8px'; 
background: `linear-gradient(-45deg, ${colors.p5} 8px, transparent 0)`;
// 45deg to flip it to the lower left corner
```

**Review by trevdor — CHANGES_REQUESTED** [review]
Thanks for building this! I love the concept -- this is gonna be a cool feature.

But to land it in the app, we'll need to tweak the approach a bit. I tried to be detailed in my suggestions, but let me know if anything I've requested is unclear.

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 10]
```diff
@@ -788,11 +856,22 @@ class BudgetGroups extends Component {
...
+  useEffect(() => {
+    const handleResize = () => {
+      setSmallVPWidth(window.outerWidth <= 365);
+    };
+
+    window.addEventListener('resize', handleResize);
+    return () => {
+      window.removeEventListener('resize', handleResize);
+    };
+  }, []);
```
[the remark this answers is not in the document]
Comment (Crazypkr1099) [reply packages/desktop-client/src/components/budget/MobileBudgetTable.js]: I didn't know it was this easy to get the width boolean! Thank you for this!

File: packages/desktop-client/src/components/budget/MobileBudgetTable.js [hunk 11]
```diff
@@ -815,11 +834,39 @@ export function BudgetTable(props) {
...
+  function toggleDisplay() {
+    setShowBudgetedCol(!showBudgetedCol);
+    if (!showBudgetedCol) {
+      // remember the pref indefinitely
+      dispatch(savePrefs({ mobileShowBudgetedColPref: true }));
+    }
+  }
+  let buttonStyle = {
+    padding: 0,
+    backgroundColor: 'transparent',
```
Comment (trevdor) [inline packages/desktop-client/src/components/budget/MobileBudgetTable.js:859]: The corner triangle gets a little cut off with the button radius.
```suggestion
    backgroundColor: 'transparent',
    borderRadius: 'unset',
```

**Review by trevdor — CHANGES_REQUESTED** [review]
@Crazypkr1099 You took a load of feedback in stride. Impressively quick turnaround. I think it's looking great! Left one small change suggestion and then I'm good.

**Comment by Crazypkr1099** [comment]
> @Crazypkr1099 You took a load of feedback in stride. Impressively quick turnaround. I think it's looking great! Left one small change suggestion and then I'm good.

Great suggestion I totally missed that!

Thanks can't wait to see this added.

**Review by trevdor — APPROVED** [review]
Thanks again!

**Comment by MatissJanis** [comment]
Sadly I can't push the new screenshot myself to your repo as I don't have the access. But here it is. Can you please upload it in place of `Mobile-loads-the-budget-page-with-budgeted-amounts-1-chromium-linux.png`?

Mobile-loads-the-budget-page-with-budgeted-amounts-1-chromium-linux

**Review by MatissJanis — APPROVED** [review]
Thanks!
